django/utils/text.py

```diff
@@ -103,38 +110,46 @@ def handle_endtag(self, tag):
             self.tags.remove(tag)
 
     def handle_data(self, data):
-        if self.func == "words":
-            data = re.split(r"(?<=\S)\s(?=\S)", data)
-            output = escape(" ".join(data[: self.num]))
-        else:
-            output = escape("".join(data[: self.num]))
-
+        data, output = self.process(data)
```

Comment:
We can get rid of the whole `func` equal to `"chars"` or `"words"` thing by subclassing and moving this logic into a separate method.